### ISSUE TITLE
[FAISS] Add explicit reusable index sessions

### ISSUE THREAD

**Issue by huguesva** [TorchDR/TorchDR#302]
## Motivation

IVF/PQ training and database ingestion can dominate repeated neighbor searches. TorchDR currently treats FAISS indexes as temporary details of a distance call, preventing explicit reuse across affinity construction, parameter sweeps, and out-of-sample operations.

## Proposed scope

- Introduce a small index/session lifecycle: configure, train if needed, add, search repeatedly, close.
- Keep the first implementation internal or experimental; do not force estimators to manage raw FAISS objects.
- Allow an estimator or affinity to accept either precomputed `(distances, indices)` or a validated TorchDR-owned FAISS session.
- Validate dimension, metric, dtype, device/distribution context, sample count, and relevant configuration before reuse.
- Make invalidation explicit. Do not infer cache validity from Python tensor identity because tensors can be mutated.
- Define deterministic cleanup and prevent live GPU resources from being pickled.
- Avoid a global implicit cache in the first implementation.

## Acceptance criteria

- Repeated search reuses a built index and produces the same result as rebuilding it.
- Incompatible input/configuration reuse fails with an actionable message.
- Cleanup releases GPU index memory and is safe when called more than once.
- Existing estimator behavior requires no session object.
- Tests cover Flat, one trainable index, CPU/GPU capability paths, reuse, invalidation, and cleanup.

## Benchmark requirement

Compare rebuild-every-call with reuse over 1, 2, 5, and 10 searches. Report one-time build cost, amortized latency, and retained GPU memory. Keep the public/experimental surface only if reuse provides a material benefit for a realistic TorchDR workflow.

## Non-goals

- Automatic caching based on tensor identity.
- On-disk index persistence.
- Exposing arbitrary native FAISS index objects throughout TorchDR.

## Pull request boundary

One focused PR for an internal or experimental reusable lifecycle. Do not update `RELEASES.rst`.

**Comment by huguesva** [comment]
**Verdict: REFUTE** an explicit reusable index/session API for #302, on the benchmark evidence below.

The exact-Flat default — the roadmap's mandated primary path — gets no material benefit from index reuse (build is a sub-millisecond GPU `add`). The only real reuse win is expert IVF/PQ, whose auto-presets were already refuted in #304 and which only pays off after many repeated searches on the *same* database. The session lifecycle the issue specifies (configure / train-if-needed / add / search* / close, cross-context validation, explicit invalidation, deterministic GPU cleanup, no-pickle-GPU) is not justified by the measured benefit, and a live session has to *retain* the index plus FAISS temp scratch between calls — reintroducing exactly the training-set memory overhead that TorchDR's `transform` path deliberately avoids by re-passing `X_train` instead of storing it.

### Setup
Single B200; `torchdr` staged node-local (`torchdr.__file__` asserted under the staged tree); clustered Gaussian data (50 blobs), d=128, k=15, sqeuclidean, seed 0, warmup + median over 5 reps. **rebuild** = the shipped `pairwise_distances_faiss` (create + train-if-needed + add + search + discard — the current per-call behaviour). **reuse** = build the index once via TorchDR's own `_create_index` / `_train_index` / `_training_sample`, then re-search. Out-of-sample model: a fixed query batch (q=10k) searched N times against a fixed database of n rows — the strongest realistic reuse case, since a self-kNN `fit` searches its database exactly once. Reuse returned **byte-identical neighbours** (`idx_eq=True`, `|ΔD|max=0`) in every case.

### Table 1 — one-time build vs per-search cost; reuse over N searches
| index | n | build (train / add) | per search | rebuild ×1 | reuse 10× | resident index |
|---|---|---|---|---|---|---|
| **Flat** (exact, default) | 100k | 0.3 ms (0 / 0.3) | 9.7 ms | 10.1 ms | **1.03×** | ~0.05 GB |
| **Flat** (exact, default) | 1M | 0.8 ms (0 / 0.8) | 93.6 ms | 94.5 ms | **1.00×** | ~0.5 GB |
| IVF (expert, approx) | 100k | 148.7 ms (112 / 36) | 0.5 ms | 149.0 ms | 9.66× | ~0.1 GB |
| IVF (expert, approx) | 1M | 2369 ms (1476 / 997) | 0.9 ms | 2533 ms | 10.65× | ~0.9 GB |

For exact Flat the "build" is just the GPU `add` (0.3–0.8 ms, i.e. 0.3–0.9 % of one search), so reuse over 10 searches saves 3–8 ms out of 0.1–0.9 s → **1.00–1.03×**. The entire reuse win is IVF/PQ *training* amortisation; a reused IVF index needs ≈16 repeated same-DB searches at 100k just to beat rebuilding an exact Flat index every call (`148.7 + 0.5·N < 10.1·N ⇒ N ≥ 16`). "resident index" is the on-GPU database copy a live session must hold for its lifetime; on top of it FAISS reserves its default `temp_memory` (~1.5 GB) per open session.

### Table 2 — the one workflow that needs a *live* index: out-of-sample `transform`
`transform(X_new, X_train)` (UMAP / LargeVis / InfoTSNE) is the only TorchDR workflow that searches a fixed database with a *different* query each call, so a reused index there would skip rebuilding `X_train`'s index. UMAP, exact-Flat default, `max_iter=300` (→ the full 100 transform optimisation steps), median of 3:

| n_train | n_new | transform() | kNN over X_train | index build (reuse skips this) | reuse 10× (saved) |
|---|---|---|---|---|---|
| 100k | 1k | 122.9 ms | 1.5 ms (1.2 %) | 0.3 ms (0.2 %) | 1.00× (0.2 %) |
| 100k | 10k | 130.8 ms | 10.1 ms (7.7 %) | 0.3 ms (0.2 %) | 1.00× (0.2 %) |
| 1M | 10k | 216.9 ms | 94.6 ms (43.6 %) | 0.8 ms (0.4 %) | 1.00× (0.3 %) |

Even where the kNN is **43.6 %** of a transform (1M database), the part a reused index can skip — the *build* — is **0.4 %**, because an exact Flat build is a sub-millisecond `add`; the kNN cost is almost entirely the *search*, which every new query pays regardless of reuse. So repeated transforms against a fixed `X_train` save 0.2–0.3 % from a reused Flat index. An IVF index would amortise its training (Table 1), but IVF is expert/approximate and #304-refuted; wiring it through this UMAP path currently also errors on an approximate index's `-1` sentinels (`torch.bincount` in the sparse affinity), a separate rough edge that argues for one exact path rather than more approximate machinery.

### Why the other repeat patterns don't need a session
Hyperparameter sweeps that re-`fit` the same `X`, and `fit`-then-eval on the same `X`, re-issue the **same** `(distances, indices)` — those are served by a precomputed-`(distances, indices)` input (the lighter alternative the issue itself lists), not a stateful GPU session; and per #304 the kNN is a negligible fraction of a DR fit at n ≤ 100k, so even that saves little end-to-end.

### Measured / not measured
Measured on 1×B200 up to n=1M: build vs search vs rebuild, reuse totals over 1/2/5/10 searches, retained GPU memory, exact-neighbour correctness of reuse, and the `transform`-workflow dilution. Not measured: multi-GPU/distributed sessions (the roadmap keeps 1 process/GPU and a session adds nothing to the collective search path), on-disk persistence (a stated non-goal), and PQ/`fast` (which #304 already showed has no usable operating point at embedding dims).

### Recommendation
Close #302 as won't-implement for a general session API. Keep the single exact path plus the existing expert `FaissConfig` override. If a concrete user need for repeated same-DB search appears, the minimal honest step is a precomputed-`(distances, indices)` input — far smaller than a GPU-resource-holding session, and it covers the sweep/eval repeat patterns — but the numbers above put even that below the materiality bar for the exact default.

### Reproduce
Build-once vs rebuild-every-call on the shipped path (standalone, no private harness):
```python
import time, torch
from torchdr.distance.faiss import (
    FaissConfig, pairwise_distances_faiss,
    _create_index, _train_index, _training_sample,
    _index_input_device, _index_device_id, faiss_device_scope,
)

dev = torch.device("cuda"); d, n, q, k = 128, 1_000_000, 10_000, 15
c = torch.randn(50, d, device=dev) * 10
DB = c[torch.randint(0, 50, (n,), device=dev)] + torch.randn(n, d, device=dev)
Q = DB[torch.randperm(n, device=dev)[:q]].clone()

def bench(index_type):
    cfg = FaissConfig(index_type=index_type)
    idev, gpu = _index_input_device(cfg, dev)
    fdev = torch.device("cuda", _index_device_id(cfg)) if gpu else None
    DBf = DB.to(idev, torch.float32).contiguous(); Qf = Q.to(idev, torch.float32).contiguous()
    def build():
        ix = _create_index("sqeuclidean", cfg, d, n, gpu)
        with faiss_device_scope(fdev):
            if not ix.is_trained:
                ix = _train_index(ix, lambda: _training_sample(DBf, 256*ix.nlist), cfg, d, gpu, None)
            ix.add(DBf)
        return ix
    def sync(): torch.cuda.synchronize(); return time.perf_counter()
    ix = build()                                             # warm
    for _ in range(2): ix.search(Qf, k); pairwise_distances_faiss(Q, k, Y=DB, config=cfg)
    t0 = sync(); [ix.search(Qf, k) for _ in range(10)]; t_reuse = sync() - t0
    t0 = sync(); [pairwise_distances_faiss(Q, k, Y=DB, config=cfg) for _ in range(10)]; t_rebuild = sync() - t0
    print(index_type, f"reuse10={t_reuse*1e3:.0f}ms rebuild10={t_rebuild*1e3:.0f}ms speedup={t_rebuild/t_reuse:.2f}x")

bench("Flat"); bench("IVF")
```

**Comment by huguesva** [comment]
Closing as not planned based on the benchmark evidence above: exact Flat index construction is negligible (about 0.2–0.4% of realistic transform time), while a stateful session would add lifecycle, validation, retained-memory, and cleanup complexity. Reopen if a concrete repeated-search workload demonstrates a material benefit that clears that maintenance cost.
